Discourage reusing Tweens

KoBeWi opened the pull request:
Closes #74245
Hopefully this will discourage people from even thinking about this 🙄

KoBeWi committed:
Discourage reusing Tweens

seppoday commented on doc/classes/Tween.xml:
```diff
@@ -89,8 +89,8 @@
 		[url=https://raw.githubusercontent.com/godotengine/godot-docs/master/img/tween_cheatsheet.png]Tween easing and transition types cheatsheet[/url]
-		[b]Note:[/b] All [Tween]s will automatically start by default. To prevent a [Tween] from autostarting, you can call [method stop] immediately after it is created.
-		[b]Note:[/b] [Tween]s are processing after all of nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
+		[b]Note:[/b] Tweens are not supposed to be reused. Create a new Tween for for each animation and every time you replay an animation from start. Also Tweens start immediately, so only create a Tween when you want to start animating. While reusing is partially supported by utilizing [method stop], it is [b]very[/b] discouraged and also pointless, as there are no performance benefits.
```
Double "for"
`Create a new Tween for for each animation and every...`

follower replied:
Perhaps it's for for for emphasis? :)

reptofrog commented on doc/classes/Tween.xml:
```diff
@@ -89,8 +89,8 @@
 		[url=https://raw.githubusercontent.com/godotengine/godot-docs/master/img/tween_cheatsheet.png]Tween easing and transition types cheatsheet[/url]
-		[b]Note:[/b] All [Tween]s will automatically start by default. To prevent a [Tween] from autostarting, you can call [method stop] immediately after it is created.
-		[b]Note:[/b] [Tween]s are processing after all of nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
+		[b]Note:[/b] Tweens are not supposed to be reused. Create a new Tween for for each animation and every time you replay an animation from start. Also Tweens start immediately, so only create a Tween when you want to start animating. While reusing is partially supported by utilizing [method stop], it is [b]very[/b] discouraged and also pointless, as there are no performance benefits.
```
I think "pointless" is bit of a strong word, the UI probably shouldn't have an opinion on what's "pointless" and what's not, it's probably best to let the user decide that.

KoBeWi replied:
Users are not the ones to decide in this case 🙂
I'm ok with rewording it, but the point is that there are no real benefits.

Mickeon replied:
I believe something much closer to the original description, while still heavily noting that reusing them is silly could be suitable.

follower commented:
A pointer for anyone else looking for more context re: `Tween`/`SceneTreeTween` + `pause()`/`stop()` code/behaviour, see:

 * https://github.com/godotengine/godot/pull/71218
 * https://github.com/godotengine/godot/commit/f64dd732b8ac249d052519940e46129960e0d771

KoBeWi commented:
Ok I reworked the PR. Instead of saying "this is somewhat supported, but don't do it", I removed all documentation notes that imply reusing (pretending it's not supported at all) and only left a comment that it results in undefined behavior.

bend-n commented on doc/classes/Tween.xml:
```diff
@@ -89,8 +89,8 @@
 		[url=https://raw.githubusercontent.com/godotengine/godot-docs/master/img/tween_cheatsheet.png]Tween easing and transition types cheatsheet[/url]
-		[b]Note:[/b] All [Tween]s will automatically start by default. To prevent a [Tween] from autostarting, you can call [method stop] immediately after it is created.
-		[b]Note:[/b] [Tween]s are processing after all of nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
+		[b]Note:[/b] Tweens are not designed to be re-used and trying to do so results in an undefined behavior. Create a new Tween for each animation and every time you replay an animation from start. Also Tweens start immediately, so only create a Tween when you want to start animating.
```
```suggestion
		[b]Note:[/b] Tweens are not designed to be re-used and trying to do so results in undefined behavior. Create a new Tween for each animation and start the animation immediately. Tweens start animating immediately, so only create a Tween when you want to start animating.
```

KoBeWi replied:
> Create a new Tween for each animation and start the animation immediately.

This repeats in the next sentence and is misleading, as it's not the user who is starting the animation.

YuriSizov replied:
Yeah, I don't understand the point of the suggested change.

bend-n replied:
the point is "*in an* undefined behaviour" is strange grammar

YuriSizov replied:
Okay, but you also removed a huge chunk of the note.

bend-n replied:
I did reword it, I just merged it with the repeated sentence after it.

YuriSizov replied:
Your suggestion removes important, if a bit verbose, information. I see no point in merging the two sentences and losing the intended message in the process for both of them.

bend-n commented on doc/classes/Tween.xml:
```diff
@@ -89,8 +89,8 @@
-		[b]Note:[/b] All [Tween]s will automatically start by default. To prevent a [Tween] from autostarting, you can call [method stop] immediately after it is created.
-		[b]Note:[/b] [Tween]s are processing after all of nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
+		[b]Note:[/b] Tweens are not designed to be re-used and trying to do so results in an undefined behavior. Create a new Tween for each animation and every time you replay an animation from start. Also Tweens start immediately, so only create a Tween when you want to start animating.
+		[b]Note:[/b] Tweens are processing after all of nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
```
```suggestion
		[b]Note:[/b] Tweens are processed after all of the nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
```

bend-n commented on doc/classes/SceneTree.xml:
```diff
@@ -88,7 +88,7 @@
 			<return type="Tween" />
 			<description>
-				Creates and returns a new [Tween].
+				Creates and returns a new [Tween]. The Tween will start automatically on the next process frame or physics frame (depending on [enum TweenProcessMode]).
```
```suggestion
				Creates and returns a new [Tween]. This Tween will start automatically on the next process or physics frame, depending on [enum TweenProcessMode].
```

bend-n commented on doc/classes/Node.xml:
```diff
@@ -188,6 +188,7 @@
 				GetTree().CreateTween().BindNode(this);
 				[/csharp]
 				[/codeblocks]
+				The Tween will start automatically on the next process frame or physics frame (depending on [enum TweenProcessMode]).
```
```suggestion
				This Tween will start automatically on the next process or physics frame, depending on [enum TweenProcessMode].
```

bend-n reviewed (approved):
just saw some english problems

YuriSizov commented on doc/classes/Tween.xml:
```diff
@@ -89,8 +89,8 @@
 		[url=https://raw.githubusercontent.com/godotengine/godot-docs/master/img/tween_cheatsheet.png]Tween easing and transition types cheatsheet[/url]
-		[b]Note:[/b] All [Tween]s will automatically start by default. To prevent a [Tween] from autostarting, you can call [method stop] immediately after it is created.
-		[b]Note:[/b] [Tween]s are processing after all of nodes in the current frame, i.e. after [method Node._process] or [method Node._physics_process] (depending on [enum TweenProcessMode]).
+		[b]Note:[/b] Tweens are not designed to be re-used and trying to do so results in an undefined behavior. Create a new Tween for each animation and every time you replay an animation from start. Also Tweens start immediately, so only create a Tween when you want to start animating.
```
```suggestion
		[b]Note:[/b] Tweens are not designed to be re-used and trying to do so results in an undefined behavior. Create a new Tween for each animation and every time you replay an animation from start. Keep in mind that Tweens start immediately, so only create a Tween when you want to start animating.
```

YuriSizov reviewed (approved):
Overall, this version seems good to me.

akien-mga commented:
Thanks!

naturally-intelligent commented:
What are best practices for persistent Tweens?

https://godotengine.org/qa/148332/best-practices-for-creating-a-persistent-tween-in-godot-4

I'm kinda stuck because if we are supposed to use AnimationPlayer for this, not sure how to dynamically add Animations at runtime either:

https://godotengine.org/qa/148336/how-can-i-add-animations-to-animationplayer-by-gdscript

KoBeWi commented:
The answer is in the docs, although probably not obvious for this particular question:
```GDScript
var tween
func animate():
	if tween:
		tween.kill() # Abort the previous animation.
		tween = create_tween()
```
Create the Tween inside `play_effect()` instead of `_ready()`. Still store it in a member variable, but use `kill()` on the old Tween (note that you don't need to check if it's still running).

YuriSizov commented:
Cherry-picked for 4.0.1.
